File: disc_golf_app/lib/widgets/skeleton_overlay.dart
import 'package:flutter/material.dart';
import '../models/form_analysis.dart';

/// Clean skeleton overlay — color-coded limbs, minimal labels.
/// In interactive mode, adjustable joints are highlighted for drag-to-correct,
/// and their border color reflects ML Kit detection confidence:
///   cyan   → confident (≥ 0.7)
///   orange → uncertain (0.5–0.7)
///   red    → unreliable (< 0.5) or unknown
class SkeletonOverlay extends CustomPainter {
  final FormAnalysis analysis;
  final int currentFrame;
  final bool interactive;
  final String? selectedLandmark;
  final Set<int>? correctedFrames;

  /// The 12 landmarks that can be manually adjusted in correction mode.
  static const adjustableLandmarks = [
    'PoseLandmarkType.rightShoulder',
    'PoseLandmarkType.leftShoulder',
    'PoseLandmarkType.rightElbow',
    'PoseLandmarkType.leftElbow',
    'PoseLandmarkType.rightWrist',
    'PoseLandmarkType.leftWrist',
    'PoseLandmarkType.rightHip',
    'PoseLandmarkType.leftHip',
    'PoseLandmarkType.rightKnee',
    'PoseLandmarkType.leftKnee',
    'PoseLandmarkType.rightAnkle',
    'PoseLandmarkType.leftAnkle',
  ];

  SkeletonOverlay({
    required this.analysis,
    required this.currentFrame,
    this.interactive = false,
    this.selectedLandmark,
    this.correctedFrames,
  });

  // Ideal angles for disc golf form
  static const _idealAngles = {
    'rightElbowAngle': 120.0,
    'leftElbowAngle': 140.0,
    'rightShoulderAngle': 100.0,
    'leftShoulderAngle': 100.0,
    'rightKneeAngle': 160.0,
    'leftKneeAngle': 160.0,
    'spineAngle': 85.0,
  };

  @override
  void paint(Canvas canvas, Size size) {
    if (analysis.frames.isEmpty) return;

    final frameIndex = currentFrame.clamp(0, analysis.frames.length - 1);
    final frame = analysis.frames[frameIndex];

    if (frame.keyPoints.isEmpty) return;

    _drawSkeleton(canvas, size, frame);
    _drawJoints(canvas, size, frame);
    _drawKeyAngleLabels(canvas, size, frame);
  }

  Offset _scalePoint(Offset point, Size canvasSize, FormFrame frame) {
    final imgW = frame.imageWidth;
    final imgH = frame.imageHeight;

    if (imgW != null && imgH != null && imgW > 0 && imgH > 0) {
      return Offset(
        point.dx * canvasSize.width / imgW,
        point.dy * canvasSize.height / imgH,
      );
    }

    if (point.dx <= 1.0 && point.dy <= 1.0) {
      return Offset(point.dx * canvasSize.width, point.dy * canvasSize.height);
    }

    return Offset(
      point.dx * canvasSize.width / 640,
      point.dy * canvasSize.height / 640,
    );
  }

  void _drawSkeleton(Canvas canvas, Size size, FormFrame frame) {
    const limbGroups = [
      _LimbGroup(
        joints: [
          ['PoseLandmarkType.rightShoulder', 'PoseLandmarkType.rightElbow'],
          ['PoseLandmarkType.rightElbow', 'PoseLandmarkType.rightWrist'],
        ],
        angleKey: 'rightElbowAngle',
      ),
      _LimbGroup(
        joints: [
          ['PoseLandmarkType.leftShoulder', 'PoseLandmarkType.leftElbow'],
          ['PoseLandmarkType.leftElbow', 'PoseLandmarkType.leftWrist'],
        ],
        angleKey: 'leftElbowAngle',
      ),
      _LimbGroup(
        joints: [
          ['PoseLandmarkType.rightHip', 'PoseLandmarkType.rightKnee'],
          ['PoseLandmarkType.rightKnee', 'PoseLandmarkType.rightAnkle'],
        ],
        angleKey: 'rightKneeAngle',
      ),
      _LimbGroup(
        joints: [
          ['PoseLandmarkType.leftHip', 'PoseLandmarkType.leftKnee'],
          ['PoseLandmarkType.leftKnee', 'PoseLandmarkType.leftAnkle'],
        ],
        angleKey: 'leftKneeAngle',
      ),
      _LimbGroup(
        joints: [
          ['PoseLandmarkType.leftShoulder', 'PoseLandmarkType.rightShoulder'],
          ['PoseLandmarkType.leftShoulder', 'PoseLandmarkType.leftHip'],
          ['PoseLandmarkType.rightShoulder', 'PoseLandmarkType.rightHip'],
          ['PoseLandmarkType.leftHip', 'PoseLandmarkType.rightHip'],
        ],
        angleKey: 'spineAngle',
      ),
    ];

    for (final group in limbGroups) {
      final angle = frame.angles[group.angleKey];
      final color = _getAngleColor(group.angleKey, angle);

      final paint = Paint()
        ..color = color.withAlpha(200)
        ..strokeWidth = 2.5
        ..style = PaintingStyle.stroke
        ..strokeCap = StrokeCap.round;

      for (final pair in group.joints) {
        final p1 = frame.keyPoints[pair[0]];
        final p2 = frame.keyPoints[pair[1]];
        if (p1 != null && p2 != null) {
          canvas.drawLine(
            _scalePoint(p1, size, frame),
            _scalePoint(p2, size, frame),
            paint,
          );
        }
      }
    }
  }

  void _drawJoints(Canvas canvas, Size size, FormFrame frame) {
    final defaultPaint = Paint()
      ..color = Colors.white
      ..style = PaintingStyle.fill;

    final borderPaint = Paint()
      ..color = Colors.black.withAlpha(120)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 1;

    for (final entry in frame.keyPoints.entries) {
      final scaled = _scalePoint(entry.value, size, frame);

      if (interactive && adjustableLandmarks.contains(entry.key)) {
        if (entry.key == selectedLandmark) {
          // Selected joint — yellow with halo
          canvas.drawCircle(
              scaled, 12, Paint()..color = Colors.yellow.withAlpha(60));
          canvas.drawCircle(scaled, 8, Paint()..color = Colors.yellow);
          canvas.drawCircle(
              scaled,
              8,
              Paint()
                ..color = Colors.black
                ..style = PaintingStyle.stroke
                ..strokeWidth = 2);
        } else {
          // Adjustable but not selected — border color reflects confidence
          final borderColor = _getConfidenceColor(
              entry.key, frame.landmarkConf);
          canvas.drawCircle(scaled, 6, Paint()..color = Colors.white);
          canvas.drawCircle(
              scaled,
              6,
              Paint()
                ..color = borderColor
                ..style = PaintingStyle.stroke
                ..strokeWidth = 2);
        }
      } else {
        // Non-interactive joint dot
        canvas.drawCircle(scaled, 3.5, defaultPaint);
        canvas.drawCircle(scaled, 3.5, borderPaint);
      }
    }
  }

  /// Returns the confidence-indicating border color for an interactive joint.
  ///   cyan   → high confidence (≥ 0.7)
  ///   orange → medium confidence (0.5 – 0.7)
  ///   red    → low confidence (< 0.5) or no data (manually corrected = trusted)
  Color _getConfidenceColor(String key, Map<String, double> conf) {
    // Manually corrected frames have empty landmarkConf — treat as trusted.
    if (conf.isEmpty) return Colors.cyanAccent;
    final c = conf[key];
    if (c == null) return Colors.redAccent;
    if (c >= 0.7) return Colors.cyanAccent;
    if (c >= 0.5) return Colors.orange;
    return Colors.redAccent;
  }

  /// Show angle labels at 3 key joints only: throwing elbow, shoulder, lead knee.
  void _drawKeyAngleLabels(Canvas canvas, Size size, FormFrame frame) {
    const labels = [
      _AngleLabel(
        vertexKey: 'PoseLandmarkType.rightElbow',
        angleKey: 'rightElbowAngle',
        offset: Offset(10, -14),
      ),
      _AngleLabel(
        vertexKey: 'PoseLandmarkType.rightShoulder',
        angleKey: 'rightShoulderAngle',
        offset: Offset(10, -14),
      ),
      _AngleLabel(
        vertexKey: 'PoseLandmarkType.rightKnee',
        angleKey: 'rightKneeAngle',
        offset: Offset(10, -14),
      ),
    ];

    for (final label in labels) {
      final vertex = frame.keyPoints[label.vertexKey];
      final angle = frame.angles[label.angleKey];
      if (vertex == null || angle == null) continue;

      final pos = _scalePoint(vertex, size, frame) + label.offset;
      final color = _getAngleColor(label.angleKey, angle);

      final text = '${angle.toStringAsFixed(0)}°';
      final textPainter = TextPainter(
        text: TextSpan(
          text: text,
          style: TextStyle(
            color: color,
            fontSize: 11,
            fontWeight: FontWeight.bold,
          ),
        ),
        textDirection: TextDirection.ltr,
      );
      textPainter.layout();

      final pillRect = RRect.fromRectAndRadius(
        Rect.fromLTWH(
          pos.dx - 3,
          pos.dy - 1,
          textPainter.width + 6,
          textPainter.height + 2,
        ),
        const Radius.circular(4),
      );

      canvas.drawRRect(
        pillRect,
        Paint()..color = Colors.black.withAlpha(180),
      );

      textPainter.paint(canvas, pos);
    }
  }

  Color _getAngleColor(String angleKey, double? angle) {
    if (angle == null) return Colors.grey;

    final ideal = _idealAngles[angleKey];
    if (ideal == null) return Colors.cyan;

    final diff = (angle - ideal).abs();
    if (diff < 10) return Colors.green;
    if (diff < 25) return Colors.orange;
    return Colors.red;
  }

  @override
  bool shouldRepaint(SkeletonOverlay oldDelegate) {
    return oldDelegate.currentFrame != currentFrame ||
        oldDelegate.selectedLandmark != selectedLandmark ||
        oldDelegate.interactive != interactive;
  }

  /// Find the nearest adjustable landmark to a canvas-space tap point.
  /// Returns the landmark key, or null if none within [threshold] pixels.
  static String? nearestLandmark(
    Offset tapPoint,
    Size canvasSize,
    FormFrame frame, {
    double threshold = 30.0,
  }) {
    String? closest;
    double closestDist = threshold;

    for (final key in adjustableLandmarks) {
      final point = frame.keyPoints[key];
      if (point == null) continue;

      final imgW = frame.imageWidth;
      final imgH = frame.imageHeight;
      Offset scaled;
      if (imgW != null && imgH != null && imgW > 0 && imgH > 0) {
        scaled = Offset(
          point.dx * canvasSize.width / imgW,
          point.dy * canvasSize.height / imgH,
        );
      } else if (point.dx <= 1.0 && point.dy <= 1.0) {
        scaled = Offset(
            point.dx * canvasSize.width, point.dy * canvasSize.height);
      } else {
        scaled = Offset(
          point.dx * canvasSize.width / 640,
          point.dy * canvasSize.height / 640,
        );
      }

      final dist = (scaled - tapPoint).distance;
      if (dist < closestDist) {
        closestDist = dist;
        closest = key;
      }
    }

    return closest;
  }

  /// Convert canvas-space coordinates back to image pixel coordinates.
  static Offset canvasToImage(
      Offset canvasPoint, Size canvasSize, FormFrame frame) {
    final imgW = frame.imageWidth;
    final imgH = frame.imageHeight;
    if (imgW != null && imgH != null && imgW > 0 && imgH > 0) {
      return Offset(
        canvasPoint.dx * imgW / canvasSize.width,
        canvasPoint.dy * imgH / canvasSize.height,
      );
    }
    return Offset(
      canvasPoint.dx / canvasSize.width,
      canvasPoint.dy / canvasSize.height,
    );
  }
}

class _LimbGroup {
  final List<List<String>> joints;
  final String angleKey;

  const _LimbGroup({required this.joints, required this.angleKey});
}

class _AngleLabel {
  final String vertexKey;
  final String angleKey;
  final Offset offset;

  const _AngleLabel({
    required this.vertexKey,
    required this.angleKey,
    required this.offset,
  });
}

File: disc_golf_android/app/src/main/kotlin/com/discflightschool/app/ui/components/SkeletonOverlay.kt
package com.discflightschool.app.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.unit.dp
import com.discflightschool.app.ui.theme.AppColors
import com.discflightschool.core.geometry.Vec2
import com.discflightschool.core.model.FormFrame
import kotlin.math.abs

/**
 * The pose skeleton drawn over a video frame.
 *
 * Limbs are coloured by how far their joint angle is from the disc golf ideal,
 * so a glance says which part of the throw is off. In correction mode the
 * adjustable joints are ringed by how much the detector trusted them — a
 * landmark it was unsure of is exactly the one worth dragging.
 */
object SkeletonOverlay {

    /** The 12 landmarks that can be adjusted by hand. */
    val ADJUSTABLE_LANDMARKS = listOf(
        "PoseLandmarkType.rightShoulder",
        "PoseLandmarkType.leftShoulder",
        "PoseLandmarkType.rightElbow",
        "PoseLandmarkType.leftElbow",
        "PoseLandmarkType.rightWrist",
        "PoseLandmarkType.leftWrist",
        "PoseLandmarkType.rightHip",
        "PoseLandmarkType.leftHip",
        "PoseLandmarkType.rightKnee",
        "PoseLandmarkType.leftKnee",
        "PoseLandmarkType.rightAnkle",
        "PoseLandmarkType.leftAnkle",
    )

    /** Target angles for a sound disc golf throw, in degrees. */
    private val IDEAL_ANGLES = mapOf(
        "rightElbowAngle" to 120.0,
        "leftElbowAngle" to 140.0,
        "rightShoulderAngle" to 100.0,
        "leftShoulderAngle" to 100.0,
        "rightKneeAngle" to 160.0,
        "leftKneeAngle" to 160.0,
        "spineAngle" to 85.0,
    )

    private data class LimbGroup(val bones: List<Pair<String, String>>, val angleKey: String)

    private val LIMB_GROUPS = listOf(
        LimbGroup(
            bones = listOf(
                "PoseLandmarkType.rightShoulder" to "PoseLandmarkType.rightElbow",
                "PoseLandmarkType.rightElbow" to "PoseLandmarkType.rightWrist",
            ),
            angleKey = "rightElbowAngle",
        ),
        LimbGroup(
            bones = listOf(
                "PoseLandmarkType.leftShoulder" to "PoseLandmarkType.leftElbow",
                "PoseLandmarkType.leftElbow" to "PoseLandmarkType.leftWrist",
            ),
            angleKey = "leftElbowAngle",
        ),
        LimbGroup(
            bones = listOf(
                "PoseLandmarkType.rightHip" to "PoseLandmarkType.rightKnee",
                "PoseLandmarkType.rightKnee" to "PoseLandmarkType.rightAnkle",
            ),
            angleKey = "rightKneeAngle",
        ),
        LimbGroup(
            bones = listOf(
                "PoseLandmarkType.leftHip" to "PoseLandmarkType.leftKnee",
                "PoseLandmarkType.leftKnee" to "PoseLandmarkType.leftAnkle",
            ),
            angleKey = "leftKneeAngle",
        ),
        LimbGroup(
            bones = listOf(
                "PoseLandmarkType.leftShoulder" to "PoseLandmarkType.rightShoulder",
                "PoseLandmarkType.leftShoulder" to "PoseLandmarkType.leftHip",
                "PoseLandmarkType.rightShoulder" to "PoseLandmarkType.rightHip",
                "PoseLandmarkType.leftHip" to "PoseLandmarkType.rightHip",
            ),
            angleKey = "spineAngle",
        ),
    )

    private data class AngleLabel(val vertexKey: String, val angleKey: String)

    /** Labels only on the throwing elbow, throwing shoulder, and lead knee. */
    private val ANGLE_LABELS = listOf(
        AngleLabel("PoseLandmarkType.rightElbow", "rightElbowAngle"),
        AngleLabel("PoseLandmarkType.rightShoulder", "rightShoulderAngle"),
        AngleLabel("PoseLandmarkType.rightKnee", "rightKneeAngle"),
    )

    /**
     * Map a landmark into canvas space.
     *
     * Landmarks arrive in whatever space the detector produced: pixels when the
     * frame's dimensions are known, normalized 0-1 for a hand-corrected frame,
     * and pixels against a 640-wide working frame when neither is recorded.
     */
    fun scalePoint(point: Vec2, canvasSize: Size, frame: FormFrame): Offset {
        val imageWidth = frame.imageWidth
        val imageHeight = frame.imageHeight

        if (imageWidth != null && imageHeight != null && imageWidth > 0 && imageHeight > 0) {
            return Offset(
                (point.x * canvasSize.width / imageWidth).toFloat(),
                (point.y * canvasSize.height / imageHeight).toFloat(),
            )
        }
        if (point.x <= 1.0 && point.y <= 1.0) {
            return Offset(
                (point.x * canvasSize.width).toFloat(),
                (point.y * canvasSize.height).toFloat(),
            )
        }
        return Offset(
            (point.x * canvasSize.width / 640).toFloat(),
            (point.y * canvasSize.height / 640).toFloat(),
        )
    }

    /** Convert canvas coordinates back into the frame's own space. */
    fun canvasToImage(canvasPoint: Offset, canvasSize: Size, frame: FormFrame): Vec2 {
        val imageWidth = frame.imageWidth
        val imageHeight = frame.imageHeight
        if (imageWidth != null && imageHeight != null && imageWidth > 0 && imageHeight > 0) {
            return Vec2(
                canvasPoint.x * imageWidth / canvasSize.width,
                canvasPoint.y * imageHeight / canvasSize.height,
            )
        }
        return Vec2(
            (canvasPoint.x / canvasSize.width).toDouble(),
            (canvasPoint.y / canvasSize.height).toDouble(),
        )
    }

    /**
     * The adjustable landmark nearest a tap, or null when nothing is within
     * [thresholdPx].
     */
    fun nearestLandmark(
        tapPoint: Offset,
        canvasSize: Size,
        frame: FormFrame,
        thresholdPx: Float = 60f,
    ): String? {
        var closest: String? = null
        var closestDistance = thresholdPx

        for (key in ADJUSTABLE_LANDMARKS) {
            val point = frame.keyPoints[key] ?: continue
            val distance = (scalePoint(point, canvasSize, frame) - tapPoint).getDistance()
            if (distance < closestDistance) {
                closestDistance = distance
                closest = key
            }
        }
        return closest
    }

    /** Green through red, by how far an angle sits from its ideal. */
    fun angleColor(angleKey: String, angle: Double?): Color {
        if (angle == null) return AppColors.Muted
        val ideal = IDEAL_ANGLES[angleKey] ?: return AppColors.Accent
        val difference = abs(angle - ideal)
        return when {
            difference < 10 -> AppColors.Good
            difference < 25 -> AppColors.Warning
            else -> AppColors.Bad
        }
    }

    /**
     * The ring colour for an adjustable joint, by detector confidence.
     *
     * A frame with no confidence data was corrected by hand, and a correction
     * the user made is the most trustworthy landmark on screen.
     */
    fun confidenceColor(key: String, confidence: Map<String, Double>): Color {
        if (confidence.isEmpty()) return AppColors.Accent
        val value = confidence[key] ?: return AppColors.Bad
        return when {
            value >= 0.7 -> AppColors.Accent
            value >= 0.5 -> AppColors.Warning
            else -> AppColors.Bad
        }
    }

    /** Draw the skeleton for [frame] across the whole draw scope. */
    fun DrawScope.drawSkeleton(
        frame: FormFrame,
        interactive: Boolean = false,
        selectedLandmark: String? = null,
    ) {
        if (frame.keyPoints.isEmpty()) return

        for (group in LIMB_GROUPS) {
            val color = angleColor(group.angleKey, frame.angles[group.angleKey])
            for ((startKey, endKey) in group.bones) {
                val start = frame.keyPoints[startKey] ?: continue
                val end = frame.keyPoints[endKey] ?: continue
                drawLine(
                    color = color.copy(alpha = 0.8f),
                    start = scalePoint(start, size, frame),
                    end = scalePoint(end, size, frame),
                    strokeWidth = 2.5.dp.toPx(),
                    cap = androidx.compose.ui.graphics.StrokeCap.Round,
                )
            }
        }

        for ((key, point) in frame.keyPoints) {
            val center = scalePoint(point, size, frame)
            if (interactive && key in ADJUSTABLE_LANDMARKS) {
                if (key == selectedLandmark) {
                    drawCircle(Color.Yellow.copy(alpha = 0.25f), radius = 12.dp.toPx(), center = center)
                    drawCircle(Color.Yellow, radius = 8.dp.toPx(), center = center)
                    drawCircle(
                        Color.Black,
                        radius = 8.dp.toPx(),
                        center = center,
                        style = Stroke(width = 2.dp.toPx()),
                    )
                } else {
                    drawCircle(Color.White, radius = 6.dp.toPx(), center = center)
                    drawCircle(
                        confidenceColor(key, frame.landmarkConf),
                        radius = 6.dp.toPx(),
                        center = center,
                        style = Stroke(width = 2.dp.toPx()),
                    )
                }
            } else {
                drawCircle(Color.White, radius = 3.5.dp.toPx(), center = center)
                drawCircle(
                    Color.Black.copy(alpha = 0.5f),
                    radius = 3.5.dp.toPx(),
                    center = center,
                    style = Stroke(width = 1.dp.toPx()),
                )
            }
        }

        drawAngleLabels(frame)
    }

    private fun DrawScope.drawAngleLabels(frame: FormFrame) {
        val textSize = 11.dp.toPx()

        for (label in ANGLE_LABELS) {
            val vertex = frame.keyPoints[label.vertexKey] ?: continue
            val angle = frame.angles[label.angleKey] ?: continue

            val position = scalePoint(vertex, size, frame) + Offset(10f, -14f)
            val color = angleColor(label.angleKey, angle)
            val text = "${angle.toInt()}°"

            val paint = android.graphics.Paint().apply {
                this.color = color.toArgb()
                this.textSize = textSize
                isAntiAlias = true
                typeface = android.graphics.Typeface.DEFAULT_BOLD
            }
            val textWidth = paint.measureText(text)

            drawRoundRect(
                color = Color.Black.copy(alpha = 0.7f),
                topLeft = Offset(position.x - 3f, position.y - textSize),
                size = Size(textWidth + 6f, textSize + 6f),
                cornerRadius = androidx.compose.ui.geometry.CornerRadius(4f, 4f),
            )

            drawContext.canvas.nativeCanvas.drawText(text, position.x, position.y, paint)
        }
    }
}

/** The skeleton for one frame, sized to fill its box. */
@Composable
fun SkeletonCanvas(
    frame: FormFrame,
    modifier: Modifier = Modifier,
    interactive: Boolean = false,
    selectedLandmark: String? = null,
) {
    Canvas(modifier) {
        with(SkeletonOverlay) {
            drawSkeleton(
                frame = frame,
                interactive = interactive,
                selectedLandmark = selectedLandmark,
            )
        }
    }
}
